### PR TITLE
Register callsites and subscribers globally

### LINKED ISSUE #61: Improve span callsite cache
For each span, a subscriber should be able to indicate:

* Always interested
* Never interested
* Sometimes interested

When a subscriber indicates that it is never interested in a span, the instrumentation overhead should not be measurable.

"Sometimes interested" supports the case of sampled instrumentation. The subscriber should be able to efficiently be able to sample traces.

### PR DESCRIPTION
**Pull request by hawkw**
Closes #61

This branch adds a global registry that tracks callsites and
subscribers. When a new callsite is registered, all subscribers are
informed of it immediately; similarly, when a new subscriber is
registered, it is informed of (and allowed to filter) all callsites.
This is used to update the filtering logic for callsites, and also
allows an opportunity for subscribers to observe a callsite's metadata
in a function that's guaranteed to be called exactly once. 

Exactly-once registration can be useful as a time where field keys 
can be extracted from the metadata and stored, ensuring that the
more costly by-name key lookup isn't repeated.

I've added a new benchmark for creating a span wuith the same callsite
multiple times, as this is where this branch offers the largest
performance improvement. 

As the becnhmark results below show, performance in that case
(`span_repeatedly`) is about 1/4th faster after this change.

Benchmarks before:

```
     Running target/release/deps/no_subscriber-30ffe0f7fbecf1ec

running 4 tests
test bench_1_atomic_load              ... bench:           0 ns/iter (+/- 0)
test bench_costly_field_no_subscriber ... bench:          48 ns/iter (+/- 9)
test bench_no_span_no_subscriber      ... bench:           0 ns/iter (+/- 0)
test bench_span_no_subscriber         ... bench:          44 ns/iter (+/- 13)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured; 0 filtered out

     Running target/release/deps/subscriber-ab6525f0578fb401

running 4 tests
test span_no_fields            ... bench:         103 ns/iter (+/- 10)
test span_repeatedly           ... bench:      21,876 ns/iter (+/- 9,254)
test span_with_fields          ... bench:         278 ns/iter (+/- 175)
test span_with_fields_add_data ... bench:       1,026 ns/iter (+/- 163)
```

Benchmarks after:
```
     Running target/release/deps/no_subscriber-0122e236a67454e8

running 4 tests
test bench_1_atomic_load              ... bench:           0 ns/iter (+/- 0)
test bench_costly_field_no_subscriber ... bench:          37 ns/iter (+/- 7)
test bench_no_span_no_subscriber      ... bench:           0 ns/iter (+/- 0)
test bench_span_no_subscriber         ... bench:          33 ns/iter (+/- 5)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured; 0 filtered out

     Running target/release/deps/subscriber-28d74558d25f84dc

running 4 tests
test span_no_fields            ... bench:          72 ns/iter (+/- 12)
test span_repeatedly           ... bench:      15,395 ns/iter (+/- 1,092)
test span_with_fields          ... bench:         155 ns/iter (+/- 14)
test span_with_fields_add_data ... bench:         698 ns/iter (+/- 87)
```

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

### CODE REVIEW

**Comment by davidbarsky** [comment]
Error on CI:

```
error[E0432]: unresolved import `tower_service::NewService`
 --> /home/travis/.cargo/git/checkouts/tower-h2-aaf68b3c8982de4c/4e1a50a/src/client/connect.rs:4:5
  |
4 | use tower_service::NewService;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^ no `NewService` in the root. Did you mean to use `Service`?
error[E0432]: unresolved import `tower_service::NewService`
 --> /home/travis/.cargo/git/checkouts/tower-h2-aaf68b3c8982de4c/4e1a50a/src/server/mod.rs:4:21
  |
4 | use tower_service::{NewService, Service};
  |                     ^^^^^^^^^^ no `NewService` in the root. Did you mean to use `Service`?
```

**Comment by hawkw** [comment]
@davidbarsky sorry about the CI breakage, that was due to an upstream dependency change. I _thought_ fd21e4bdff2c611fc46b2b7d7d39da70e5f3ba3e would be sufficient to fix it, but it wasn't --- 5b55938a4988dc8cc5972b0140678c7f35c2c7f1 should actually resolve the problem.

Ah, the perils of git dependencies!

File: tokio-trace-core/Cargo.toml [hunk 1]
```diff
@@ -7,5 +7,8 @@ authors = ["Eliza Weisman <eliza@buoyant.io>"]
 default = []
 test-support = []
 
+[dependencies]
+lazy_static = "1"
```
Comment (carllerche) [inline tokio-trace-core/Cargo.toml]: We should pin to a minor release in order to not unexpectedly have the min rustc version bumped.

Comment (hawkw) [reply tokio-trace-core/Cargo.toml]: 👍

File: tokio-trace-core/src/subscriber.rs [hunk 2]
```diff
@@ -188,6 +210,30 @@ pub trait Subscriber {
     fn close(&self, span: SpanId);
 }
 
+/// Indicates a `Subscriber`'s interest in a particular callsite.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub enum Interest {
```
[the remark this answers is not in the document]
Comment (hawkw) [reply tokio-trace-core/src/subscriber.rs]: @carllerche I was planning to do a global pass turning _all_ enums into structs prior to releasing 0.1, but can do that in this branch instead if you prefer.

[the remark this answers is not in the document]
Comment (hawkw) [reply tokio-trace-core/src/subscriber.rs]: @carllerche fixed in ec8299bf9a77b59a3d803a89629b8c8392239112.

**Pull request merged**